File: Remove Duplicates from Sorted Array II.cpp
#define F(i, a, b) for(int i = a; i <= b; i++)
#include <vector>
#include <unordered_map>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    int removeDuplicates(int A[], int n) {
        if( n < 3 ){
            return n;
        }

        unordered_map<int, int> ht;

        int len = n;
        for( int i = 0, index = 0; i <= n-1; i++, index++ ){
            if( ht.find(A[i]) != ht.end() ){
                if( ht[A[i]] == 1 ){
                    ht[A[i]] = 2;
                    if( index < i ){
                        A[index] = A[i];
                    }
                }else{
                    //delete
                    index--;
                    len--;
                }
            }else{
                ht[A[i]] = 1;
                if( index < i ){
                    A[index] = A[i];
                }
            }
        }

        return len;
    }
};

File: LeetCode-Java/Remove Duplicates from Sorted Array II.java
public class Solution {
	public int removeDuplicates(int[] A) {
		int len = A.length;
		if( len < 3 ){
			return len;
		}
		int index = 1;
		int processing = A[0];
		boolean isTwice = false;
		for( int i = 1; i < A.length; i++ ){
			if( A[i] == processing ){
				if( isTwice ){
					len--;
				}else{
					isTwice = true;
					if( index < i ){
						A[index] = A[i];
					}
					index++;
				}
			}else{
				processing = A[i];
				isTwice = false;
				if( index < i ){
					A[index] = A[i];
				}
				index++;
			}
		}
		return len;
	}
}
